mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -780,7 +718,78 @@ impl S3CrtClientInner {
             parse_meta_request_error,
             move |result| _ = tx.send(result.map(|_| std::mem::take(&mut *body.lock().unwrap()))),
         )?;
-        Ok(S3HttpRequest {
+        Ok(S3MetaRequest {
+            receiver: rx.fuse(),
+            meta_request,
+        })
+    }
+
+    /// Make a meta-request using this S3 client that returns the response headers on success or
+    /// invokes the given callback on failure.
+    ///
+    /// The `parse_meta_request_error` callback is invoked on failed requests. It should return `None`
+    /// if it doesn't have a request-specific failure reason. The client will apply some generic error
+    /// parsing in this case (e.g. for permissions errors).
+    fn meta_request_with_headers_payload<E: std::error::Error + Send + 'static>(
+        &self,
+        options: MetaRequestOptions,
+        request_span: Span,
+        parse_meta_request_error: impl FnOnce(&MetaRequestResult) -> Option<E> + Send + 'static,
+    ) -> Result<S3MetaRequest<Headers, E>, S3RequestError> {
+        let (tx, rx) = oneshot::channel::<ObjectClientResult<Headers, E, S3RequestError>>();
+
+        // On success, stash the headers in this lock during the on_headers callback,
+        // and pull them out during the on_meta_request_result callback.
+        let on_headers: Arc<Mutex<Option<Headers>>> = Default::default();
+        let on_result = on_headers.clone();
+
+        let meta_request = self.meta_request_with_callbacks(
+            options,
+            request_span,
+            |_| {},
+            move |headers, status| {
+                // Only store headers if we have a 2xx status code. If we only get other status codes,
+                // then on_meta_request_result will send an error.
+                if (200..300).contains(&status) {
+                    *on_headers.lock().unwrap() = Some(headers.clone());
+                }
+            },
+            |_, _| {},
+            parse_meta_request_error,
+            move |result| {
+                // On success, extract the headers.
+                _ = tx.send(result.map(|_| on_result.lock().unwrap().take().unwrap()))
```

Comment:
Same here about unwrapping - is it better than logging and returning some kind of internal error?

Reply:
Yes. Fixed that.